util/retry: TestRetryWithMaxDuration failed

util/retry.TestRetryWithMaxDuration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20557888?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20557888?buildTab=artifacts#/) on release-25.4 @ [0af2a332521ffbc302e71faff74fe112a81cc35e](https://github.com/cockroachdb/cockroach/commits/0af2a332521ffbc302e71faff74fe112a81cc35e):


```
=== RUN   TestRetryWithMaxDuration
--- FAIL: TestRetryWithMaxDuration (0.00s)
=== RUN   TestRetryWithMaxDuration/errors_with_opt.Closer_that_is_closed
    retry_test.go:554: 
        	Error Trace:	pkg/util/retry/retry_test.go:554
        	Error:      	"23ms" is not less than or equal to "20ms"
        	Test:       	TestRetryWithMaxDuration/errors_with_opt.Closer_that_is_closed
        	Messages:   	expected time spent to be less than or equal to max expected time spent
--- FAIL: TestRetryWithMaxDuration/errors_with_opt.Closer_that_is_closed (0.00s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/unowned
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetryWithMaxDuration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55107

--- a/pkg/util/retry/BUILD.bazel
+++ b/pkg/util/retry/BUILD.bazel
@@ -25,7 +25,6 @@ go_test(
     ],
     embed = [":retry"],
     deps = [
-        "//pkg/testutils/skip",
         "//pkg/util/log",
         "//pkg/util/timeutil",
         "@com_github_cockroachdb_errors//:errors",

--- a/pkg/util/retry/retry.go
+++ b/pkg/util/retry/retry.go
@@ -172,6 +172,15 @@ func (r *Retry) Next() bool {
 		return false
 	}
 
+	// Check for cancellation first to prioritize over timer.
+	select {
+	case <-r.opts.Closer:
+		return false
+	case <-r.ctx.Done():
+		return false
+	default:
+	}
+
 	backoff, actualWait, shouldAttempt := r.calcDurationScopedBackoff()
 
 	if !shouldAttempt && r.opts.PreemptivelyCancel {

--- a/pkg/util/retry/retry_test.go
+++ b/pkg/util/retry/retry_test.go
@@ -10,7 +10,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
@@ -402,11 +401,7 @@ func TestRetryWithMaxDuration(t *testing.T) {
 		retryFunc         func(attemptNum int) error
 		preRetryFunc      func()
 		expectedTimeSpent time.Duration
-		// For cases where the amount of time spent is not deterministic, we can set
-		// an upper bound instead (e.g. for context or closer).
-		maxExpectedTimeSpent time.Duration
-		expectedErr          bool
-		skipUnderDuress      bool
+		expectedErr       bool
 	}
 
 	testCases := []testcase{
@@ -474,11 +469,8 @@ func TestRetryWithMaxDuration(t *testing.T) {
 			preRetryFunc: func() {
 				cancelCtxFunc()
 			},
-			maxExpectedTimeSpent: time.Millisecond * 20,
-			expectedErr:          true,
-			// Under duress, closing a context will not necessarily stop the retry
-			// loop immediately, so we skip this test under duress.
-			skipUnderDuress: true,
+			expectedTimeSpent: 0,
+			expectedErr:       true,
 		},
 		{
 			name: "errors with opt.Closer that is closed",
@@ -494,20 +486,13 @@ func TestRetryWithMaxDuration(t *testing.T) {
 			preRetryFunc: func() {
 				close(closeCh)
 			},
-			maxExpectedTimeSpent: time.Millisecond * 20,
-			expectedErr:          true,
-			// Under duress, closing a channel will not necessarily stop the retry
-			// loop immediately, so we skip this test under duress.
-			skipUnderDuress: true,
+			expectedTimeSpent: 0,
+			expectedErr:       true,
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			if tc.skipUnderDuress && skip.Duress() {
-				skip.UnderDuress(t, "skipping test under duress: %s", tc.name)
-			}
-
 			timeSource := timeutil.NewManualTime(time.Now())
 			tc.opts.Clock = timeSource
 			// Disable randomization for deterministic tests.
@@ -544,18 +529,9 @@ func TestRetryWithMaxDuration(t *testing.T) {
 				require.NoError(t, err)
 			}
 
-			if tc.expectedTimeSpent != 0 {
-				require.Equal(
-					t, tc.expectedTimeSpent, timeSource.Since(start), "expected time does not match actual spent time",
-				)
-			}
-
-			if tc.maxExpectedTimeSpent != 0 {
-				require.LessOrEqual(
-					t, timeSource.Since(start), tc.maxExpectedTimeSpent,
-					"expected time spent to be less than or equal to max expected time spent",
-				)
-			}
+			require.Equal(
+				t, tc.expectedTimeSpent, timeSource.Since(start), "expected time does not match actual spent time",
+			)
 		})
 	}
 }

